File: src/agents/reporter.ts
import { getSupabaseClient } from '@/lib/supabase';
import { logger } from '@/lib/logger';
import type { DailyReportInsert, NewsArticleRow, RiskLevel, ArticleCategory } from '@/types';

// ── 카운터 헬퍼 ───────────────────────────────────────────────────────────────

function countBy<T>(items: T[], key: keyof T): Map<string, number> {
  const map = new Map<string, number>();
  for (const item of items) {
    const v = String(item[key] ?? 'unknown');
    map.set(v, (map.get(v) ?? 0) + 1);
  }
  return map;
}

function topN(counter: Map<string, number>, n: number): string[] {
  return [...counter.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, n)
    .map(([k]) => k);
}

// ── 마크다운 빌더 ─────────────────────────────────────────────────────────────

function buildMarkdown(
  reportDate: string,
  articles: NewsArticleRow[],
  counts: { total: number; critical: number; high: number; medium: number; low: number },
  topDiseases: string[],
  topKeywords: string[],
  categoryStats: Map<string, number>
): string {
  const now = new Date().toISOString().replace('T', ' ').slice(0, 16) + ' UTC';
  const riskEmoji: Record<RiskLevel, string> = {
    critical: '🔴', high: '🟠', medium: '🟡', low: '🟢', unknown: '⚪',
  };

  const lines: string[] = [
    `# 🏥 의료 뉴스 일일 리포트 — ${reportDate}`,
    `> 생성: ${now} | 총 ${counts.total}건`,
    '',
    '## 위험도 요약',
    `| 위급 🔴 | 고위험 🟠 | 중위험 🟡 | 저위험 🟢 | 합계 |`,
    `|:------:|:-------:|:-------:|:-------:|:----:|`,
    `| ${counts.critical} | ${counts.high} | ${counts.medium} | ${counts.low} | ${counts.total} |`,
    '',
    '## 카테고리별 분포',
    ...[...categoryStats.entries()]
      .sort((a, b) => b[1] - a[1])
      .map(([cat, cnt]) => `- **${cat}**: ${cnt}건`),
    '',
    '## 주요 질병·병원체',
    topDiseases.length > 0 ? topDiseases.map((d) => `\`${d}\``).join(' · ') : '해당 없음',
    '',
    '## 핵심 키워드',
    topKeywords.length > 0 ? topKeywords.map((k) => `\`${k}\``).join(' · ') : '해당 없음',
  ];

  const urgent = articles.filter((a) => a.risk_level === 'critical' || a.risk_level === 'high');
  if (urgent.length > 0) {
    lines.push('', '## ⚠️ 주요 뉴스 (위급·고위험)');
    for (const a of urgent.slice(0, 20)) {
      const emoji = riskEmoji[a.risk_level as RiskLevel] ?? '⚪';
      lines.push(
        '',
        `### ${emoji} ${a.title}`,
        `**출처**: ${a.source} | **분류**: ${a.category ?? '기타'} | **위험도**: ${a.risk_level}`,
      );
      if (a.summary) lines.push('', a.summary);
      if (a.diseases?.length > 0) lines.push('', `**질병명**: ${a.diseases.join(', ')}`);
      lines.push('', `🔗 ${a.url}`);
    }
  }

  lines.push('', '## 전체 기사 목록');
  for (const a of articles.slice(0, 100)) {
    const emoji = riskEmoji[a.risk_level as RiskLevel] ?? '⚪';
    lines.push(`- ${emoji} [${a.title}](${a.url}) — ${a.source} · ${a.category ?? '기타'}`);
  }

  return lines.join('\n');
}

// ── 메인 ──────────────────────────────────────────────────────────────────────

export async function generateDailyReport(targetDate?: string): Promise<DailyReportInsert | null> {
  const reportDate = targetDate ?? new Date().toISOString().slice(0, 10);
  const startISO = `${reportDate}T00:00:00+00:00`;
  const endISO = `${reportDate}T23:59:59+00:00`;

  logger.info(`[Reporter] ${reportDate} 리포트 생성 중...`);

  let articles: NewsArticleRow[];
  try {
    const db = getSupabaseClient();
    const { data, error } = await db
      .from('news_articles')
      .select('*')
      .gte('collected_at', startISO)
      .lte('collected_at', endISO)
      .order('risk_level', { ascending: true });

    if (error) throw error;
    articles = (data ?? []) as NewsArticleRow[];
  } catch (e) {
    logger.error(`[Reporter] DB 조회 오류: ${e}`);
    return null;
  }

  if (articles.length === 0) {
    logger.info(`[Reporter] ${reportDate} 기사 없음`);
    return null;
  }

  const riskCounter = countBy(articles, 'risk_level');
  const categoryCounter = countBy(articles, 'category');

  const allDiseases = articles.flatMap((a) => a.diseases ?? []);
  const allKeywords = articles.flatMap((a) => a.keywords ?? []);

  const diseaseCounter = new Map<string, number>();
  for (const d of allDiseases) diseaseCounter.set(d, (diseaseCounter.get(d) ?? 0) + 1);

  const keywordCounter = new Map<string, number>();
  for (const k of allKeywords) keywordCounter.set(k, (keywordCounter.get(k) ?? 0) + 1);

  const topDiseases = topN(diseaseCounter, 15);
  const topKeywords = topN(keywordCounter, 15);

  const counts = {
    total: articles.length,
    critical: riskCounter.get('critical') ?? 0,
    high: riskCounter.get('high') ?? 0,
    medium: riskCounter.get('medium') ?? 0,
    low: riskCounter.get('low') ?? 0,
  };

  const markdown = buildMarkdown(reportDate, articles, counts, topDiseases, topKeywords, categoryCounter);

  const report: DailyReportInsert = {
    report_date: reportDate,
    total_articles: counts.total,
    critical_count: counts.critical,
    high_count: counts.high,
    medium_count: counts.medium,
    low_count: counts.low,
    top_diseases: topDiseases,
    top_keywords: topKeywords,
    report_markdown: markdown,
    report_json: {
      date: reportDate,
      counts,
      top_diseases: topDiseases,
      top_keywords: topKeywords,
      category_stats: Object.fromEntries(categoryCounter),
      articles: articles.map((a) => ({
        title: a.title, url: a.url, source: a.source,
        risk_level: a.risk_level, category: a.category,
        summary: a.summary, diseases: a.diseases,
      })),
    },
  };

  try {
    const db = getSupabaseClient();
    const { error } = await db
      .from('daily_reports')
      .upsert(report, { onConflict: 'report_date' });
    if (error) throw error;
    logger.info(`[Reporter] 저장 완료: ${reportDate}, ${counts.total}건`);
  } catch (e) {
    logger.error(`[Reporter] DB 저장 오류: ${e}`);
  }

  return report;
}

File: src/agents/reporter.py
import json
from datetime import date, datetime, timezone, timedelta
from collections import Counter
from src.db.client import get_client
from src.db.models import DailyReport
from src.utils.logger import logger


def _build_markdown(report_date: str, articles: list, counts: dict, top_diseases: list, top_keywords: list) -> str:
    lines = [
        f"# 의료 뉴스 일일 리포트 — {report_date}",
        f"\n> 생성 시각: {datetime.now(timezone.utc).strftime('%Y-%m-%d %H:%M UTC')}",
        "\n## 요약 통계",
        f"- 총 수집 기사: **{counts['total']}건**",
        f"- 🔴 위급(Critical): {counts['critical']}건",
        f"- 🟠 고위험(High): {counts['high']}건",
        f"- 🟡 중위험(Medium): {counts['medium']}건",
        f"- 🟢 저위험(Low): {counts['low']}건",
        "\n## 주요 질병/병원체",
        ", ".join(top_diseases) if top_diseases else "해당 없음",
        "\n## 핵심 키워드",
        ", ".join(top_keywords) if top_keywords else "해당 없음",
    ]

    # 위급/고위험 기사 목록
    critical_articles = [a for a in articles if a.get("risk_level") in ("critical", "high")]
    if critical_articles:
        lines.append("\n## 주요 뉴스 (위급·고위험)")
        for a in critical_articles[:20]:
            risk_emoji = "🔴" if a.get("risk_level") == "critical" else "🟠"
            lines.append(f"\n### {risk_emoji} {a.get('title', '')}")
            lines.append(f"**출처**: {a.get('source', '')} | **분류**: {a.get('category', '')}")
            if a.get("summary"):
                lines.append(f"\n{a['summary']}")
            if a.get("diseases"):
                lines.append(f"\n**질병명**: {', '.join(a['diseases'])}")
            lines.append(f"\n🔗 [{a.get('url', '')}]({a.get('url', '')})")

    # 전체 기사 목록 (최대 50건)
    lines.append("\n## 전체 기사 목록")
    for a in articles[:50]:
        risk = a.get("risk_level", "unknown")
        emoji = {"critical": "🔴", "high": "🟠", "medium": "🟡", "low": "🟢"}.get(risk, "⚪")
        lines.append(f"- {emoji} [{a.get('title','')}]({a.get('url','')}) — {a.get('source','')}")

    return "\n".join(lines)


async def generate_daily_report(target_date: date | None = None) -> dict:
    """당일 수집 기사로 일일 리포트 생성 후 DB 저장."""
    if target_date is None:
        target_date = date.today()

    report_date_str = target_date.isoformat()
    start_dt = datetime.combine(target_date, datetime.min.time()).replace(tzinfo=timezone.utc)
    end_dt = start_dt + timedelta(days=1)

    logger.info(f"[Reporter] {report_date_str} 리포트 생성 중...")

    try:
        db = get_client()
        res = db.table("news_articles") \
            .select("*") \
            .gte("collected_at", start_dt.isoformat()) \
            .lt("collected_at", end_dt.isoformat()) \
            .order("risk_level") \
            .execute()
        articles = res.data or []
    except Exception as e:
        logger.error(f"[Reporter] DB 조회 오류: {e}")
        return {}

    if not articles:
        logger.info(f"[Reporter] {report_date_str} 기사 없음")
        return {}

    risk_counter = Counter(a.get("risk_level", "unknown") for a in articles)
    disease_counter = Counter(d for a in articles for d in (a.get("diseases") or []))
    keyword_counter = Counter(k for a in articles for k in (a.get("keywords") or []))

    top_diseases = [d for d, _ in disease_counter.most_common(15)]
    top_keywords = [k for k, _ in keyword_counter.most_common(15)]

    counts = {
        "total": len(articles),
        "critical": risk_counter.get("critical", 0),
        "high": risk_counter.get("high", 0),
        "medium": risk_counter.get("medium", 0),
        "low": risk_counter.get("low", 0),
    }

    markdown = _build_markdown(report_date_str, articles, counts, top_diseases, top_keywords)
    report_json = {
        "date": report_date_str,
        "counts": counts,
        "top_diseases": top_diseases,
        "top_keywords": top_keywords,
        "articles": [
            {
                "title": a.get("title"),
                "url": a.get("url"),
                "source": a.get("source"),
                "risk_level": a.get("risk_level"),
                "category": a.get("category"),
                "summary": a.get("summary"),
                "diseases": a.get("diseases"),
                "keywords": a.get("keywords"),
            }
            for a in articles
        ],
    }

    report = DailyReport(
        report_date=report_date_str,
        **counts,
        top_diseases=top_diseases,
        top_keywords=top_keywords,
        report_markdown=markdown,
        report_json=report_json,
    )

    try:
        db = get_client()
        db.table("daily_reports").upsert(
            report.model_dump(mode="json"),
            on_conflict="report_date",
        ).execute()
        logger.info(f"[Reporter] 리포트 저장 완료: {report_date_str}, {counts['total']}건")
    except Exception as e:
        logger.error(f"[Reporter] 리포트 DB 저장 오류: {e}")

    # 마크다운 파일로도 저장
    import os
    os.makedirs("reports", exist_ok=True)
    with open(f"reports/{report_date_str}.md", "w", encoding="utf-8") as f:
        f.write(markdown)
    logger.info(f"[Reporter] 파일 저장: reports/{report_date_str}.md")

    return report_json
